### PR TITLE
Debug sync current token

### PR DESCRIPTION
**Pull request by Fizzadar**
Something I threw together to aid debugging sync issues so we can pull a sync response between two points in time, rather than always since -> now. Thought I'd upstream if interested, feel free to close :)

Signed off by Nick @ Beeper (@fizzadar).

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog).
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: changelog.d/14047.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Add debug parameter to sync API to override the current token. Contributed by Nick @ Beeper (@fizzadar).
```
Comment (MadLittleMods) [inline changelog.d/14047.misc:1]:
```suggestion
Add `?synapse_debug_current_token` query parameter to `/sync` API to override the current token. Contributed by Nick @ Beeper (@fizzadar).
```

[the remark this answers is not in the document]
Comment (Fizzadar) [reply changelog.d/14047.misc]: Yeah I thought this too, but thought out of scope to make that change, although seems like replacing `now_token` with `current_token` would be a sensible consistency improvement 👍

File: synapse/rest/client/sync.py [hunk 2]
```diff
@@ -175,6 +175,11 @@ async def on_GET(self, request: SynapseRequest) -> Tuple[int, JsonDict]:
         if since is not None:
             since_token = await StreamToken.from_string(self.store, since)
 
+        debug_token = parse_string(request, "synapse_debug_current_token")
```
Comment (MadLittleMods) [inline synapse/rest/client/sync.py:178]: We could document the `?synapse_debug_current_token` query parameter under a new "Debugging" page under the "Development" section

File: synapse/handlers/sync.py [hunk 3]
```diff
@@ -288,6 +288,7 @@ async def wait_for_sync_for_user(
         since_token: Optional[StreamToken] = None,
```
Comment (MadLittleMods) [inline synapse/handlers/sync.py:288]: Some discussion in `#synapse-dev` about this PR: https://matrix.to/#/!vcyiEtMVHIhWXcJAfl:sw1v.org/$oaV7wGmCQOD-pByNgH6x1YhrVrqHBnEuKD76QQ7bClU?via=matrix.org&via=element.io&via=beeper.com

**Comment by clokep** [comment]
I think the general feeling of the time is that we wouldn't want to merge this into mainline, but having some better way to debug sync weirdness is a good thing (maybe refactoring so this is easier via the manhole or something?)

Overall I think this PR in its current form is not something we're willing to accept.

(If others disagree, do feel free to shout and re-open.)

**Pull request closed without merging**